## Predicting Shots Made Per Game by Kobe Bryant

*Predict how many shots Kobe Bryant made per game*

**Dataset:** https://www.kaggle.com/c/kobe-bryant-shot-selection/data

**Approach:**

Exploring regularization techniques on linear regression model
- RidgeCV 
- LassoCV
- Elastic Net



### I. Load packages and data.

In [1]:
import numpy as np
import pandas as pd
import patsy

from sklearn.linear_model import Ridge, Lasso, ElasticNet, LinearRegression, RidgeCV, LassoCV, ElasticNetCV
from sklearn.model_selection import cross_val_score

import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
plt.style.use('fivethirtyeight')

%config InlineBackend.figure_format = 'retina'
%matplotlib inline

In [2]:
kobe = pd.read_csv('./datasets/kobe_superwide_games.csv')

In [3]:
kobe.head()

SHOTS_MADE  AWAY_GAME  SEASON_OPPONENT:atl:1996-97  \
0         0.0          0                          0.0   
1         0.0          1                          0.0   
2         2.0          1                          0.0   
3         2.0          1                          0.0   
4         0.0          0                          1.0   

   SEASON_OPPONENT:atl:1997-98  SEASON_OPPONENT:atl:1999-00  \
0                          0.0                          0.0   
1                          0.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   

   SEASON_OPPONENT:atl:2000-01  SEASON_OPPONENT:atl:2001-02  \
0                          0.0                          0.0   
1                          0.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   

   SEASON_OPPONENT:atl:2002-03  SEASON_OPPONENT:atl:2003-04  \
0                          0.0                          0.0   
1                          0.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   

   SEASON_OPPONENT:atl:2004-05         ...          \
0                          0.0         ...           
1                          0.0         ...           
2                          0.0         ...           
3                          0.0         ...           
4                          0.0         ...           

   ACTION_TYPE:tip_layup_shot  ACTION_TYPE:tip_shot  \
0                         0.0                   0.0   
1                         0.0                   0.0   
2                         0.0                   0.0   
3                         0.0                   0.0   
4                         0.0                   0.0   

   ACTION_TYPE:turnaround_bank_shot  \
0                               0.0   
1                               0.0   
2                               0.0   
3                               0.0   
4                               0.0   

   ACTION_TYPE:turnaround_fadeaway_bank_jump_shot  \
0                                             0.0   
1                                             0.0   
2                                             0.0   
3                                             0.0   
4                                             0.0   

   ACTION_TYPE:turnaround_fadeaway_shot  \
0                                   0.0   
1                                   0.0   
2                                   0.0   
3                                   0.0   
4                                   0.0   

   ACTION_TYPE:turnaround_finger_roll_shot  ACTION_TYPE:turnaround_hook_shot  \
0                                      0.0                               0.0   
1                                      0.0                               0.0   
2                                      0.0                               0.0   
3                                      0.0                               0.0   
4                                      0.0                               0.0   

   ACTION_TYPE:turnaround_jump_shot  SEASON_GAME_NUMBER  CAREER_GAME_NUMBER  
0                               0.0                   1                   1  
1                               0.0                   2                   2  
2                               0.0                   3                   3  
3                               0.0                   4                   4  
4                               0.0                   5                   5  

[5 rows x 645 columns]

In [4]:
kobe.shape

(1558, 645)

### II. Examining the data.

In [5]:
print('Columns:', len(kobe.columns))

Columns: 645


In [6]:
print(kobe.columns[0:20])

Index(['SHOTS_MADE', 'AWAY_GAME', 'SEASON_OPPONENT:atl:1996-97',
       'SEASON_OPPONENT:atl:1997-98', 'SEASON_OPPONENT:atl:1999-00',
       'SEASON_OPPONENT:atl:2000-01', 'SEASON_OPPONENT:atl:2001-02',
       'SEASON_OPPONENT:atl:2002-03', 'SEASON_OPPONENT:atl:2003-04',
       'SEASON_OPPONENT:atl:2004-05', 'SEASON_OPPONENT:atl:2005-06',
       'SEASON_OPPONENT:atl:2006-07', 'SEASON_OPPONENT:atl:2007-08',
       'SEASON_OPPONENT:atl:2008-09', 'SEASON_OPPONENT:atl:2009-10',
       'SEASON_OPPONENT:atl:2010-11', 'SEASON_OPPONENT:atl:2011-12',
       'SEASON_OPPONENT:atl:2012-13', 'SEASON_OPPONENT:atl:2013-14',
       'SEASON_OPPONENT:atl:2014-15'],
      dtype='object')


**Observations:**

SHOTS_MADE will be the target variable for prediction.

Feature selection will have to be performed as there are many columns (645), and many columns represent similar things, so there is the issue of multicollinearity.


### III. Create predictor and target variables. Standardize the predictors.


In [7]:
y = kobe.SHOTS_MADE.values
X = kobe.iloc[:,1:]

# Initialize the StandardScaler object.
from sklearn.preprocessing import StandardScaler

ss = StandardScaler()

# Use the "fit_transform" function to standardize the X design matrix.
Xs = ss.fit_transform(X)

### IV. Building a linear regression predicting SHOTS_MADE from the rest of the columns.

In [8]:
linreg = LinearRegression()

linreg_scores = cross_val_score(linreg, Xs, y, cv=10)

print(linreg_scores)
print(np.mean(linreg_scores))

[-1.79977947e+29 -1.56491766e+28 -6.34863163e+27 -2.62406933e+27
 -1.57635452e+28 -3.07382964e+27 -6.38281933e+27 -5.84388364e+27
 -5.19939021e+27 -3.77445313e+28]
-2.7860782364257116e+28


**Observations:**

The mean R^2 is extremely negative. 
The linear regression is performing far worse than the baseline on the testing sets.
This is probably due to overfitting.

### V. Using RidgeCV

In [9]:
ridge_alphas = np.logspace(-2, 7, 50)

optimal_ridge = RidgeCV(alphas=ridge_alphas, cv=10)
optimal_ridge.fit(Xs, y)

print(optimal_ridge.alpha_)

910.2981779915227


In [10]:
ridge = Ridge(alpha=optimal_ridge.alpha_)

ridge_scores = cross_val_score(ridge, Xs, y, cv=10)

print(ridge_scores)
print(np.mean(ridge_scores))

[0.564539   0.52759942 0.48654074 0.59889514 0.52757419 0.52897799
 0.52767583 0.41256579 0.40442795 0.44688295]
0.5025679004240269


### VI. Using LassoCV

In [11]:
optimal_lasso = LassoCV(n_alphas=500, cv=10, verbose=1)
optimal_lasso.fit(Xs, y)

print(optimal_lasso.alpha_)

........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

0.06790539979326354


In [12]:
lasso = Lasso(alpha=optimal_lasso.alpha_)

lasso_scores = cross_val_score(lasso, Xs, y, cv=10)

print(lasso_scores)
print(np.mean(lasso_scores))

[0.61603625 0.52641159 0.53888743 0.61322432 0.54627539 0.5554072
 0.52716822 0.44849563 0.4646267  0.51838952]
0.5354922262084527


In [13]:
lasso.fit(Xs, y)

Lasso(alpha=0.06790539979326354, copy_X=True, fit_intercept=True,
   max_iter=1000, normalize=False, positive=False, precompute=False,
   random_state=None, selection='cyclic', tol=0.0001, warm_start=False)

In [14]:
lasso_coefs = pd.DataFrame({'variable':X.columns,
                            'coef':lasso.coef_,
                            'abs_coef':np.abs(lasso.coef_)})

lasso_coefs.sort_values('abs_coef', inplace=True, ascending=False)

lasso_coefs.head(20)

variable      coef  abs_coef
579            COMBINED_SHOT_TYPE:jump_shot  1.233880  1.233880
574                SHOT_TYPE:2pt_field_goal  0.866201  0.866201
566         SHOT_ZONE_BASIC:restricted_area  0.453050  0.453050
577                 COMBINED_SHOT_TYPE:dunk  0.274926  0.274926
611                   ACTION_TYPE:jump_shot -0.272130  0.272130
575                SHOT_TYPE:3pt_field_goal  0.140492  0.140492
582                       SECONDS_REMAINING  0.132360  0.132360
612                  ACTION_TYPE:layup_shot -0.129658  0.129658
423             SEASON_OPPONENT:sac:1999-00  0.127065  0.127065
561       SHOT_ZONE_BASIC:above_the_break_3  0.118469  0.118469
233             SEASON_OPPONENT:mem:2009-10  0.116944  0.116944
549                          SEASON:2015-16 -0.115167  0.115167
333             SEASON_OPPONENT:nyk:2008-09  0.111245  0.111245
569                SHOT_ZONE_AREA:center(c)  0.104304  0.104304
395             SEASON_OPPONENT:phx:2009-10  0.099200  0.099200
594  ACTION_TYPE:driving_floating_jump_shot -0.096740  0.096740
466             SEASON_OPPONENT:sea:2004-05  0.094990  0.094990
38              SEASON_OPPONENT:bos:2015-16 -0.089163  0.089163
556               SHOT_ZONE_RANGE:16-24_ft.  0.079334  0.079334
456             SEASON_OPPONENT:sas:2014-15 -0.075074  0.075074

In [15]:
print('Percent variables zeroed out:', np.sum((lasso.coef_ == 0))/float(len(lasso.coef_)))

Percent variables zeroed out: 0.7888198757763976


### VII. Using ElasticNetCV

In [16]:
l1_ratios = np.linspace(0.01, 1.0, 25)

optimal_enet = ElasticNetCV(l1_ratio=l1_ratios, n_alphas=30, cv=10)
optimal_enet.fit(Xs, y)

print(optimal_enet.alpha_)
print(optimal_enet.l1_ratio_)

/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of

0.07685093298363148
0.9587500000000001


In [17]:
enet = ElasticNet(alpha=optimal_enet.alpha_, l1_ratio=optimal_enet.l1_ratio_)

enet_scores = cross_val_score(enet, Xs, y, cv=10)

print(enet_scores)
print(np.mean(enet_scores))

[0.61442115 0.52698943 0.53876714 0.61223638 0.5482063  0.55522845
 0.52397675 0.44858197 0.46393239 0.52269212]
0.5355032083264541


**Conclusion:**

RidgeCV $R^2$ score: 0.503

LassoCV $R^2$ score: 0.535

Elastic Net $R^2$ score: 0.536


We can see that the elastic net approach performs about the same as lasso, which is to be expected given the
l1_ration approaching 1 (full lasso).
